packages/source-fetcher/lib/sourcify.ts

```diff
@@ -0,0 +1,136 @@
+  implements Fetcher {
+  get fetcherName(): string {
+    return "sourcify";
+  }
+  static get fetcherName(): string {
+    return "sourcify";
+  }
+
+  static async forNetworkId(
+    id: number,
+    options?: Types.FetcherOptions
+  ): Promise<SourcifyFetcher> {
+    //in the future, we may add protocol and node options,
+    //but these don't exist yet
+    return new SourcifyFetcher(id);
+  }
+
+  private networkId: number;
+
+  private domain: string = "contractrepo.komputing.org";
+
+  constructor(networkId: number) {
+    //we'll just treat all network IDs as valid;
+    //that wouldn't work very well if we were using the IPFS interface,
+    //which goes by network name, but it'll work fine for the HTTP interface
+    //(no idea why they're different)
+    //(technically it should be chain ID but realistically it won't matter)
+    this.networkId = networkId;
+  }
+
+  async isNetworkValid(): Promise<boolean> {
+    return true;
+  }
+
+  async fetchSourcesForAddress(
+    address: string
+  ): Promise<Types.SourceInfo | null> {
+    const metadata = await this.getMetadata(address);
+    if (!metadata) {
+      return null;
+    }
+    let sources: Types.SourcesByPath;
+    if (metadata.language === "Solidity") {
+      sources = Object.assign(
+        {},
+        ...(await Promise.all(
+          Object.entries(metadata.sources).map(
+            async ([sourcePath, { content: source }]) => ({
+              [sourcePath]:
+                source !== undefined
+                  ? source //sourcify doesn't support this yet but they're planning it
+                  : await this.getSource(address, sourcePath)
+            })
+          )
+        ))
+      );
+    } else {
+      //don't bother attempting to fetch sources if it's Yul
+      sources = {};
+    }
+    return {
+      sources,
+      options: {
+        language: metadata.language,
+        version: metadata.compiler.version,
+        settings: removeLibraries(metadata.settings)
+      }
+    };
+  }
+
+  private async getMetadata(
+    address: string
+  ): Promise<Types.SolcMetadata | null> {
+    try {
+      return await this.requestWithRetries<Types.SolcMetadata>({
+        uri: `https://${this.domain}/contract/${
+          this.networkId
+        }/${address}/metadata.json`,
+        json: true //turns on auto-parsing
+      });
+    } catch (error) {
+      //is this a 404 error? if so just return null
+      if (error.statusCode === 404) {
+        return null;
+      }
+      //otherwise, we've got a problem; rethrow the error
+      throw error;
+    }
+  }
+
+  private async getSource(
+    address: string,
+    sourcePath: string
+  ): Promise<string> {
+    return await this.requestWithRetries<string>({
+      uri: `https://${this.domain}/contract/${
+        this.networkId
+      }/${address}/sources/${sourcePath}`
+    });
+  }
+
+  private async requestWithRetries<T>(
+    requestObject: any //sorry, trying to import the type properly ran into problems
```

Comment:
If you can't import this type, maybe you can create one inside this file (`requestObjectType` or some such) instead of calling this `any`? Not a big deal but might be worth considering.

Reply:
Hm, maybe?  Not sure it's worth the trouble in this case, bunch of weird little options.

Reply:
Lately I've been making a point to do what you describe @fainashalts, but because this is a `private` method, I think it's less important here than if it were public? Your call @haltman-at

Reply:
OK, gonna leave it alone then.